inventory_management_system_api/migrations/scripts/expected_lifetime_migration.py

```
@@ -0,0 +1,188 @@
+    """
+    Output database model for a catalogue item.
+    """
+
+    id: StringObjectIdField = Field(alias="_id")
+    catalogue_category_id: StringObjectIdField
+    manufacturer_id: StringObjectIdField
+    obsolete_replacement_catalogue_item_id: Optional[StringObjectIdField] = None
+    properties: List[PropertyOut] = []
+
+    model_config = ConfigDict(populate_by_name=True, arbitrary_types_allowed=True)
+
+
+# pylint: enable=duplicate-code
+
+
+class Migration(BaseMigration):
+    """Migration for Catalogue Items' Optional Expected Lifetime"""
+
+    description = "Migration for Catalogue Items' Optional Expected Lifetime"
+
+    def __init__(self, database: Database):
+
+        self._catalogue_items_collection: Collection = database.catalogue_items
+
+    def forward(self, session: ClientSession):
+        """This function should actually perform the migration
+
+        All database functions should be given the session in order to ensure all updates are done within a transaction
+        """
+        catalogue_items = self._catalogue_items_collection.find({}, session=session)
+
+        logger.info("expected_lifetime forward migration")
+        for catalogue_item in catalogue_items:
+            try:
+                old_catalogue_item = OldCatalogueItemOut(**catalogue_item)
+
+                new_catalogue_item = NewCatalogueItemIn(**old_catalogue_item.model_dump())
+
+                update_data = {
+                    **new_catalogue_item.model_dump(),
+                    "modified_time": old_catalogue_item.modified_time,
+                }
+
+                self._catalogue_items_collection.replace_one(
+                    {"_id": catalogue_item["_id"]},
+                    update_data,
+                    session=session,
+                )
+
+            except ValidationError as ve:
+                logger.error("Validation failed for item with id %s: %s", catalogue_item["_id"], ve)
+
+                continue
```

This shouldn't either shouldn't be caught here or you need to raise the exception again e.g. `raise ve` - the way the migration script works is that it runs `forward` in a transaction meaning that if an error occurs and isn't caught, then the whole migration reverts. Whereas by catching it will carry on meaning the database could end up with some valid documents migrated, and some invalid ones not.

---

Ah, okay I think I understand. The error should be uncaught, so that the transaction gets terminated, because we don't some documents with the field and others missing it. So I can either raise another error, or just remove the `except`.

---

Yes exactly